# Exploring eBay Car Sales Data
> Bộ dữ liệu eBay Car là bộ dữ liệu về việc rao vặt các ô tô đã qua sử dụng trên Kaggle , tải tại [đây](https://www.kaggle.com/datasets/sijovm/used-cars-data-from-ebay-kleinanzeigen)
> 
> Mục đích dự án: Làm sạch dữ liệu và phân tích cơ bản

In [55]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "latin1 ")

In [56]:
pd.set_option('display.max_rows', None)

# Hiển thị toàn bộ cột
pd.set_option('display.max_columns', None)

In [57]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

In [58]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

#### Ta có thể thấy rằng tên các cột đã được chuẩn hóa theo cùng định dạng nhưng chưa phải dạng Snakecase, hầu hết các cột các có đủ các giá trị, tuy nhiên còn các cột tồn tại Mising Value như vehicleType, gearBox, model, fuelType, notRepaireDamage
Các cột bị miss khoảng 7-10%, cá biệt notRepairDamage miss 25%

#  1. Làm sạch dữ liệu và phân tích cơ bản

### 1.1 Chuẩn hóa tên cột


In [59]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [60]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [61]:
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158 

Tên các cột đã được chuẩn hóa theo dạng snake_case, thích hợp hơn với Python và tăng sự nhất quán 

In [62]:
print(autos.describe(include = 'all'))


               date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
cou

> #### 1. Các cột có ít giá trị chiếm đa số tần suất như cột `seller` có giá trị private chiếm 99% tần suất, `offer_type` với giá trị Angebot, `nr_of_pictures`, đây là các cột không có quá nhiều thông tin quan trọng, đa dạng
> #### 2. Các cột cần được chuyển về dạng numberic là `price`, `odometer`
> #### 3. Các cột cần điều tra thêm là cột `registration_year` khi có năm dao động từ 1000 - 9999, chắc chắn có sai sót khi nhập liệu . Bên cạnh đó, trong cột `power_ps` có giá trị lớn nhất là 17700 ps, đây là con số chưa thể đạt được của xe ô tô, do đó cần phải xem xét lại những sai sót
> 

## 1.2 : Chuyển định dạng cột ( `price` và `odometer` )


Để thuận tiện cho việc tính toán, các cột vốn dĩ phải là thuộc kiểu dữ liệu số học như `price` và `odometer` cần phải được thay đổi lại do chúng đang ở kiểu dữ liệu chuỗi

In [63]:
print(autos["price"].unique())
print(autos["price"].dtype)
print(autos["odometer"].unique())
print(autos["odometer"].dtype)

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
object
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']
object


> Các giá trị ở cột price đang ở type object, có pattern là "$" + số tiền
> Các giá trị ở cột odometer có pattern là số km + "km"
> Tuy nhiên dấu "," ở các giá trị không thể hiện đúng, cần xóa bỏ

In [64]:
#Xóa kí hiệu $ ở các giá trị cột price và xóa kí hiệu km ở cột odometer
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")

In [65]:
# Chuyển cột price thành dạng float, odometer sang dạng int
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

In [66]:
print(autos["price"].unique())
print(autos["price"].dtype)
print(autos["odometer"].unique())
print(autos["odometer"].dtype)

[ 5000  8500  8990 ...   385 22200 16995]
int64
[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]
int64


In [67]:
# Đổi tên cột
autos.rename({"price" : "price_dollar"}, inplace = True, axis = 1)
autos.rename({"odometer":"odometer_km"}, inplace = True, axis = 1)

In [68]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price_dollar        50000 non-null  int64 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gear_box            47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

### Kiểm tra giá trị các cột `price_dollar` và `odometer_km` và xóa ngoại lệ

In [69]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts().sort_index(ascending = False))


(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: count, dtype: int64


* Cột odometer_km có giá trị dao động trong khoảng ( 5000 - 150000 ), các giá trị này là hợp lí, tuy nhiên giá trị 5000km lại có sự tăng lên đáng ngờ
* Giá trị 1500000 chiếm đa số, chứng minh đây là bộ dữ liệu chủ yếu đến thị trường xe cũ, đi lâu ngày
* Các giá trị dưới 20000 chiếm tỷ trọng nhỏ, do đó bộ dữ liệu không đại diện cho thị trường xe mới lái

In [70]:
print(autos["price_dollar"].unique().shape)
print(autos["price_dollar"].describe())
print(autos["price_dollar"]
      .value_counts()
      .sort_index(ascending = False)
      .head(20)
     )


(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64
price_dollar
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64


Trong cột `Price`, các giá trị dao động từ 0 - 100000000, ở cận phải là con số không tưởng. Bỏ qua một số loại xe có giá trị sưu tầm lớn, thì chiếc xe thương mại đắt nhất 2016 có giá khoảng 11.1 triệu đô. 

Có vẻ như ở các giá trị của `price_dollar` tăng đều đặn đến 350000, nhưng đột biến tăng lên ở 999990, do đó ta sẽ xóa bỏ những outliers này

In [71]:
autos = autos[ autos["price_dollar"].between(1,350000)]

In [72]:
print(autos["price_dollar"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollar, dtype: float64


Giá trị trung bình của `price_dollar` đã giảm từ `9000` xuống `6000`

## 1.3: Khám phá các cột ngày, tháng

#### Hiện tại chúng ta có các cột biểu thị ngày tháng sau: 
* `date_crawled`
* `last_seen`
* `ad_created`
* `registration_month`
* `registration_year`

In [73]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [74]:
print(autos['date_crawled']
      .str[:10] # Chọn ra 10 kí tự đầu tiên thể hiện cho ngày crawl data
      .value_counts(normalize=True, dropna=False)   
      .sort_index()
     )

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64


Tốc độ crawl data không ổn định, có nhiều ngày tỷ lệ nhỏ xen kẽ giữa những ngày có tỷ lệ lớn. Tuy nhiên đây không phải dữ liệu cần quan tâm, có thể bot qua 

In [75]:
print(autos['ad_created']
      .str[:10] # Chọn ra 10 kí tự đầu tiên thể hiện cho ngày tạo quảng cáo
      .value_counts(normalize=True, dropna=False)   
      .sort_index()
     )

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000082
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001483
2016-03-05    0.022897


In [76]:
print(autos['last_seen']
      .str[:10] # Chọn ra 10 kí tự đầu tiên thể hiện cho ngày tạo quảng cáo
      .value_counts(normalize=True, dropna=False)   
      .sort_index()
     )

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64


In [77]:
print(autos["registration_year"].describe())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Không có quá nhiều thông tin hữu ích từ các cột ngày tháng, tuy nhiên ở cột registration, tồn tại giá trị vô lí 1000 - 9999

## 1.4 Xóa các giá trị sai

Xóa các giá trị sai ở cột `registration_year`

Điều nổi bật tìm thấy khi phân tích dữ liệu trong cột năm đăng kí `registration_year` đó là: 
* Giá trị năm nhỏ nhất là 1000
* Giá trị năm lớn nhất là 9999
Cần lưu ý rằng đây là bộ dữ liệu về những chiếc xe cũ từ năm 2016 trờ về trước, do đó ta cần loại bỏ các cột có giá trị vô lí

In [78]:
print(autos["registration_year"]
      .value_counts()
      .sort_index()
     )

registration_year
1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
1969      19
1970      38
1971      26
1972      33
1973      23
1974      24
1975      18
1976      21
1977      22
1978      44
1979      34
1980      85
1981      28
1982      41
1983      51
1984      51
1985      95
1986      72
1987      72
1988     135
1989     174
1990     347
1991     339
1992     370
1993     425
1994     629
1995    1227
1996    1373
1997    1951
1998    2363
1999    2897
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010   

Rõ ràng, tồn tại các giá trị đột biến, lưu ý rằng xe ô tô chỉ sản xuất quy mô lớn và hoàn thiện cấu tạo ở những năm 1900, do đó ta loại bỏ các giá trị ngoại lệ trước đó và các giá trị ngoài 2016

In [79]:
autos = autos[ autos["registration_year"].between(1900, 2016)]

In [80]:
print(autos["registration_year"]
      .value_counts(normalize = True)
      .sort_values()
     )

registration_year
1931    0.000021
1943    0.000021
1939    0.000021
1952    0.000021
1938    0.000021
1948    0.000021
1927    0.000021
1953    0.000021
1929    0.000021
1934    0.000043
1957    0.000043
1954    0.000043
1951    0.000043
1941    0.000043
1955    0.000043
1950    0.000064
1958    0.000086
1937    0.000086
1962    0.000086
1956    0.000086
1910    0.000107
1959    0.000129
1961    0.000129
1963    0.000171
1964    0.000257
1965    0.000364
1975    0.000386
1969    0.000407
1976    0.000450
1977    0.000471
1966    0.000471
1973    0.000493
1960    0.000493
1974    0.000514
1968    0.000557
1971    0.000557
1967    0.000557
1981    0.000600
1972    0.000707
1979    0.000728
1970    0.000814
1982    0.000878
1978    0.000943
1984    0.001093
1983    0.001093
1987    0.001542
1986    0.001542
1980    0.001821
1985    0.002035
1988    0.002892
1989    0.003727
1991    0.007262
1990    0.007433
1992    0.007926
2015    0.008397
1993    0.009104
1994    0.013474
2014    0.014

Được rồi, các cột bây giờ đã phù hợp với thực tế hơn về ngày đăng kí 

Số lượng xe được rao bán nhiều nhất là các dòng các năm 1999 - 2011

Với ô tô, quãng đường di chuyển một năm rao động từ 10000 - 15000 km là hợp lí, với số lượng xe cũ trong cột `odometer_km` với các giá trị 50000 - 150000 km là chủ yếu thì các ròng xe năm 1999-2011 chiếm đa số trên thị trường là hợp lí

## 1.5 Khám phá giá theo hãng xe

Wow, có nhiều câu hỏi chúng ta cần quan tâm đến các hãng xe:
* Hãng xe nào đang được rao bán nhiều nhất
* Hãng xe nào có giá trung bình rẻ nhất
* Hãng xe nào có giá trị trung bình đắt nhất
Trước hết, hãy xem có bao nhiêu dòng xe đang tồn tại trong bộ dữ liệu

In [81]:
print(autos["brand"].value_counts(normalize = True ))

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

Hầu hết các hãng xe trong bộ dữ liệu đều là các hãng nổi tiếng, tuy nhiên một vài hàng chiếm số lượng xe nhiều hơn, trong khi một số lại chiếm lượng xe nhỏ

Để không cần phải quan tâm nhiều, ta chỉ chọn những hãng xe chiếm trên 1% lượng xe để tổng hợp

In [82]:
# Lấy tên các dòng xe chiếm trên 1%
brand = autos["brand"].value_counts(normalize = True )  # Tạo Series chỉ gồm tên hãng xe và tỷ lệ lượng xe
brand_over_one_percent = brand[ brand >= 0.01]    # Lấy ra Series tên hãng xe có số lượng lớn hơn 1%
list_brand = brand_over_one_percent.index.to_list()   # Tạo list gồm các index của series trên 
print(list_brand)

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart', 'citroen', 'toyota', 'hyundai']


Ta đã có tên của các hãng xe chiếm đa số trên thị trường xe cũ trong bộ dữ liệu trên. Tiếp theo ta cần trả lời câu hỏi về giá trị trung bình của các xe theo hãng

In [83]:
# Tạo dictionary lưu giá trị trung bình của các hãng xe
mean_price_of_each_brand = {}
for brand in list_brand:
    critiea = autos["brand"] == brand
    mean_price = autos.loc[critiea, "price_dollar"].mean()
    mean_price_of_each_brand[brand] = mean_price
    

In [84]:
mean_price_of_brand = pd.Series(mean_price_of_each_brand) # Chuyển dictionary sang series
print(mean_price_of_brand.sort_values())

renault          2474.864607
fiat             2813.748538
opel             2975.241935
peugeot          3094.017229
smart            3580.223903
ford             3749.469507
citroen          3779.139144
mazda            4112.596615
seat             4397.230950
nissan           4743.402525
toyota           5167.091062
hyundai          5365.254274
volkswagen       5402.410262
skoda            6368.000000
bmw              8332.820518
mercedes_benz    8628.450366
audi             9336.687454
dtype: float64


Thực tế đã được phản ánh đúng !!! 

* Những chiếc xe hãng AUDI, MERCEDES-BENZ, AUDI thực sự là các hãng xe mà có giá trị cao nhất
* Hãng nổi tiếng với giá tầm trung phù hợp với các tầng lớp trung lưu như OPEL, PEUGOET có giá trị thấp nhất
* VOLKSWAGEN, HUYNDAI, TOYOTA, NISSAN, MAZDA, FORD là các hãng xe ở giữa
Chúng ta sẽ đều bất ngờ khi những chiếc xe TOYOTA, HUYNDAI, NISSAN lại có giá trị trung bình cao hơn MAZDA và FORD


Chúng ta sẽ so sánh số năm đi trung bình và số km đã đi trung bình của các loại xe

In [85]:
# Số năm đi trung bình
mean_year_of_each_brand = {}
for brand in list_brand:
    critiea = autos["brand"] == brand
    year = 2016 - autos.loc[critiea, "registration_year"] 
    mean_year = year.mean()
    mean_year_of_each_brand[brand] = mean_year
print(pd.Series(mean_year_of_each_brand))

volkswagen       13.532752
bmw              12.964765
opel             13.776782
mercedes_benz    13.968688
audi             11.844098
ford             13.230156
renault          13.135847
peugeot          12.251256
fiat             12.782790
seat             11.525205
skoda             9.472585
nissan           12.562412
mazda            13.036671
smart            10.517398
citroen          11.541284
toyota           11.657673
hyundai           8.811966
dtype: float64


In [86]:
# Số km đi trung bình
mean_odo_of_each_brand = {}
for brand in list_brand:
    critiea = autos["brand"] == brand
    odo = autos.loc[critiea, "odometer_km"] 
    mean_odo = odo.mean()
    mean_odo_of_each_brand[brand] = mean_odo
print(pd.Series(mean_odo_of_each_brand))

volkswagen       128707.158791
bmw              132572.513140
opel             129310.035842
mercedes_benz    130788.363313
audi             129157.386785
ford             124266.012872
renault          128071.331213
peugeot          127153.625269
fiat             117121.971596
seat             121131.301290
skoda            110848.563969
nissan           118330.995792
mazda            124464.033850
smart             99326.777610
citroen          119694.189602
toyota           115944.350759
hyundai          106442.307692
dtype: float64


In [87]:
df = pd.DataFrame([pd.Series(mean_price_of_each_brand), pd.Series(mean_year_of_each_brand), pd.Series(mean_odo_of_each_brand)]).T
df.columns = ["price", "year", "odometer"]
print(df.sort_values(by = ["price", "year", "odometer"]))


                     price       year       odometer
renault        2474.864607  13.135847  128071.331213
fiat           2813.748538  12.782790  117121.971596
opel           2975.241935  13.776782  129310.035842
peugeot        3094.017229  12.251256  127153.625269
smart          3580.223903  10.517398   99326.777610
ford           3749.469507  13.230156  124266.012872
citroen        3779.139144  11.541284  119694.189602
mazda          4112.596615  13.036671  124464.033850
seat           4397.230950  11.525205  121131.301290
nissan         4743.402525  12.562412  118330.995792
toyota         5167.091062  11.657673  115944.350759
hyundai        5365.254274   8.811966  106442.307692
volkswagen     5402.410262  13.532752  128707.158791
skoda          6368.000000   9.472585  110848.563969
bmw            8332.820518  12.964765  132572.513140
mercedes_benz  8628.450366  13.968688  130788.363313
audi           9336.687454  11.844098  129157.386785


Các xe hãng Ford và Mazda đều có số năm di chuyển và số km di chuyển lớn hơn Huyndai, Toyota và Nissan, do đó giá xe thấp hơn cũng là điều đương nhiên.

Tuy Huyndai có lượng năm đi và số km di chuyển ít nhất nhưng giá không cao hơn so với các dòng Toyota, Nissan là bao nhiêu, do đó Huyndai có thể sẽ là hãng xe giá thấp hơn các hãng xe khác nếu cùng số năm, số km di chuyển

Chúng ta cũng có thể hiểu rằng đây là bộ dữ liệu được lấy từ người Đức, do đó giá những chiếc xe nội địa sẽ rẻ hơn những loại xe nhập khẩu từ nước ngoài như các hãng xe Nhật, Hàn

# 2. Làm sạch, phân tích mở rộng

### 2.1 Chuyển đổi từ tiếng Đức sang tiếng Anh

Các cột cần đổi là các cột có giá trị là các string, chỉ có từ 1 - 5 giá trị, các cột đó là `seller`,`offer_type`,`gear_box`,`unrepaired_damage`

In [88]:
print(autos["seller"].unique())
print(autos["offer_type"].unique())
print(autos["gear_box"].unique())
print(autos["unrepaired_damage"].unique())


['privat' 'gewerblich']
['Angebot']
['manuell' 'automatik' nan]
['nein' nan 'ja']


In [89]:
autos["seller"] = autos["seller"].map( {"privat" : "private", "gewerblich": "commercial"})
autos["offer_type"] = autos["offer_type"].map({'Angebot': 'offer', 'Gesuch': 'request'})
autos['gear_box'] = autos['gear_box'].map({'manuell': 'manual', 'automatik': 'automatic'})
autos['unrepaired_damage'] = autos['unrepaired_damage'].map({'ja': 'yes', 'nein': 'no'})

### 2.2 Xếp hạng các model phổ biến

In [90]:
print(autos[["brand", "model"]]
      .value_counts()
      .sort_values(ascending = False)
      .head(20)
     )

brand          model      
volkswagen     golf           3707
bmw            3er            2615
volkswagen     polo           1609
opel           corsa          1592
volkswagen     passat         1349
opel           astra          1348
audi           a4             1231
mercedes_benz  c_klasse       1136
bmw            5er            1132
mercedes_benz  e_klasse        958
audi           a3              825
               a6              797
ford           focus           762
               fiesta          722
volkswagen     transporter     674
renault        twingo          615
peugeot        2_reihe         600
smart          fortwo          550
opel           vectra          544
mercedes_benz  a_klasse        539
Name: count, dtype: int64


Volkswagen là hãng phổ biến nhất 

### 2.3 Chia odometer theo khoảng

Chúng ta phân chia số lượng theo các khoảng trong cột odometer với các khoảng : 
* < 50k
* 50k - 100k
* 100k - 150k
* `> 150k

In [101]:
# Tạo category với pandas.cut()
bins = [0,50000,100000,150000, np.inf]
label = ["<= 50k", "50k - 100k", "100k - 150k", ">150k" ]
autos["km_group"] = pd.cut(autos["odometer_km"], bins = bins, labels = label)
print(autos["km_group"]
      .value_counts()
      .sort_values(ascending = False)
     )

km_group
100k - 150k    34942
50k - 100k      7421
<= 50k          4318
>150k              0
Name: count, dtype: int64


In [120]:
# Tính giá trung bình theo lượng km đi
mean = autos.groupby("km_group", observed=True)["price_dollar"].mean().rename("avg_price").reset_index()
print(mean)

      km_group     avg_price
0       <= 50k  14890.828856
1   50k - 100k   9595.419350
2  100k - 150k   4107.938641


### 2.4 So sánh giá giữa các xe đã qua sửa chữa và chưa qua sửa chữa

In [142]:
mean_repaired = autos.loc[autos["unrepaired_damage"] == "yes", "price_dollar"].mean()
mean_unrepaired = autos.loc[autos["unrepaired_damage"] == "no", "price_dollar"].mean()
rate = mean_unrepaired / mean_repaired
print(f'Mean price of unrepaired car more {rate}x than mean price of repaired cars')

Mean price of unrepaired car more 3.196593613330288x than mean price of repaired cars


In [140]:
print(mean_unrepaired)
print(mean_repaired)

2241.146035242291
7164.033102796004


## Kết luận

Trong quá trình phân tích và làm sạch dữ liệu, chúng ta đã làm việc với bộ dữ liệu những xe đã qua sử dụng và được rao vặt trên trang eBay của Đức.
Chúng ta đã loại bỏ những giá trị không thực tế, để lại các xe được sản xuất từ 1900 - 2016, loại bỏ các xe có giá ngoại lệ tăng vọt từ 350000 $ trờ lên. 

Thực tế rằng chúng ta có thể kết luận một số thứ sau:

1. Đây là bộ dữ liệu của người Đức do đó các hãng xe Đức chiếm đa số
2. Các loại xe nhập khẩu từ nước ngoài, cho dù đối với chúng ta đó là các hãng xe nổi tiếng với giá cả tầm trung thì khi vào Đức, giá của nó vẫn không hề rẻ hơn một số hãng mà chúng ta cho là cao cấp. Các hãng xe Đức cũng có sự cách biệt nhau về giá trị, cụ thể
    * Xe của các hãng ở top đầu Là AUDI, MERCEDES-BENZ, BMW vẫn là những hãng xe mà có giá trị cao nhất khi đã qua sử dụng 8000 - 10000$
      
    * Xe của hãng VOLKSWAGEN là những xe ở tầm trung 5500$
  
      
    * Trong khi xe của MAZDA, FORD là các hãng xe có giá trị thấp hơn cả với giá trị trung bình 3500 - 4500 $
3. Các xe cũ được bán hầu hết đã được đi trên 10 - 11 năm, có số cây đi được là trên 150000km, giá trị các xe giảm dần theo số lượng cây số đi được và số năm từ ngày lăn bánh, tuy nhiên sự tương quan này chỉ giời hạn trong bộ dữ liệu được DATAQUEST cung cấp
4. Các xe chưa từng qua sử chữa có giá cao gấp 2 - 3 lẫn những xe đã trải qua sử chữa 